File: SEP3_Database/src/main/java/com/example/SEP3_Database/model/domain/list/message/GroupMessageList.java
package com.example.SEP3_Database.model.domain.list.message;

import com.example.SEP3_Database.model.domain.unit.message.GroupMessage;

import java.util.ArrayList;

public class GroupMessageList {
    private ArrayList<GroupMessage> MessageList;

    public GroupMessageList()
    {
        MessageList = new ArrayList<>();
    }

    public void addMessage(GroupMessage newMessage)
    {
        if (newMessage!=null)
        {
            MessageList.add(newMessage);
        }
    }

    public int getSize()
    {
        return MessageList.size();
    }

    public GroupMessage getMessageByIndex(int index)
    {
        if (index>=0&&index<MessageList.size())
        {
            return MessageList.get(index);
        }
        else
        {
            return null;
        }
    }

    public GroupMessageList getMessageBySenderId(String id)
    {
        var messageList = new GroupMessageList();
        for (int x=0;x<MessageList.size();x++)
        {
            if (MessageList.get(x).SenderId.equals(id))
            {
                messageList.addMessage(MessageList.get(x));
            }
        }
        return messageList;
    }

    public GroupMessageList getMessageByGroupId(String id)
    {
        var messageList = new GroupMessageList();
        for (int x=0;x<MessageList.size();x++)
        {
            if (MessageList.get(x).getGroupId().equals(id))
            {
                messageList.addMessage(MessageList.get(x));
            }
        }
        return messageList;
    }

    public void removeMessageById(String senderId, String groupId)
    {
        ArrayList<GroupMessage> groupMessageList = new ArrayList<>();
        if (senderId!=null)
        {
            if (groupId!=null)
            {
                for (int x=0;x<MessageList.size();x++)
                {
                    if (!MessageList.get(x).SenderId.equals(senderId) || !MessageList.get(x).getGroupId().equals(groupId))
                    {
                        groupMessageList.add(MessageList.get(x));
                    }
                }
            }
            else
            {
                for (int x=0;x<MessageList.size();x++)
                {
                    if (!MessageList.get(x).SenderId.equals(senderId))
                    {
                        groupMessageList.add(MessageList.get(x));
                    }
                }
            }
        }
        else
        {
            if (groupId!=null)
            {
                for (int x=0;x<MessageList.size();x++)
                {
                    if (!MessageList.get(x).getGroupId().equals(groupId))
                    {
                        groupMessageList.add(MessageList.get(x));
                    }
                }
            }
        }
        MessageList = groupMessageList;
    }

    public GroupMessageList copy()
    {
        var messageList = new GroupMessageList();
        for (int x=0;x<MessageList.size();x++)
        {
            messageList.MessageList.add(MessageList.get(x).copy());
        }
        return messageList;
    }
}

File: SEP3_Client/Model/List/Message/GroupMessageList.cs
﻿using System.Collections.Generic;
using System.Linq;
using SEP3_Client.Model.List.Group;
using SEP3_Client.Model.Unit.Message;

namespace SEP3_Client.Model.List.Message
{
    public class GroupMessageList
    {
        public List<GroupMessage> MessageList { get; set; }

        public GroupMessageList()
        {
            MessageList = new List<GroupMessage>();
        }

        public void AddMessage(GroupMessage newMessage)
        {
            MessageList.Add(newMessage);
        }

        public int GetSize()
        {
            return MessageList.Count;
        }

        public GroupMessage GetMessageByIndex(int index)
        {
            if (index>=0&&index<MessageList.Count)
            {
                return MessageList[index];
            }
            else
            {
                return null;
            }
        }

        public GroupMessageList GetMessageBySenderId(string id)
        {
            var messageList = new GroupMessageList();
            foreach (var message in MessageList)
            {
                if (message.SenderId==id)
                {
                    messageList.AddMessage(message);
                }
            }
            return messageList;
        }
        
        public GroupMessageList GetMessageByGroupId(string id)
        {
            var messageList = new GroupMessageList();
            foreach (var message in MessageList)
            {
                if (message.GroupId==id)
                {
                    messageList.AddMessage(message);
                }
            }
            return messageList;
        }
        
        public GroupMessageList GetMessageByUserId(string id)
        {
            var messageList = new GroupMessageList();
            foreach (var message in MessageList)
            {
                if (ChatGroupList.GetAllGroupList().GetGroupByGroupId(message.GroupId).HasId(id))
                {
                    messageList.AddMessage(message);
                }
            }
            return messageList;
        }
        
        public void RemoveMessageById(string senderId, string groupId)
        {
            if (!string.IsNullOrEmpty(senderId))
            {
                if (!string.IsNullOrEmpty(groupId))
                {
                    foreach (var message in MessageList.Where(message => message.SenderId==senderId&&message.GroupId==groupId))
                    {
                        MessageList.Remove(message);
                    }
                }
                else
                {
                    foreach (var message in MessageList.Where(message => message.SenderId==senderId))
                    {
                        MessageList.Remove(message);
                    }
                }
            }
            else
            {
                if (!string.IsNullOrEmpty(groupId))
                {
                    foreach (var message in MessageList.Where(message => message.GroupId==groupId))
                    {
                        MessageList.Remove(message);
                    }
                }
            }
        }
        
        public GroupMessageList Copy()
        {
            var messageList = new GroupMessageList();
            foreach (var message in MessageList)
            {
                messageList.AddMessage(message.Copy());
            }
            return messageList;
        }
    }
}